arviz/stats/stats.py

```diff
@@ -342,91 +364,139 @@ def hpd(ary, credible_interval=None, circular=False, multimodal=False, skipna=Fa
-        intervals = bins[idx][density[idx].cumsum() <= credible_interval]
-        intervals.sort()
+    min_idx = np.argmin(interval_width)
+    hdi_min = ary[min_idx]
+    hdi_max = ary[min_idx + interval_idx_inc]
 
-        intervals_splitted = np.split(intervals, np.where(np.diff(intervals) >= dx * 1.1)[0] + 1)
+    if circular:
+        hdi_min = hdi_min + mean
+        hdi_max = hdi_max + mean
+        hdi_min = np.arctan2(np.sin(hdi_min), np.cos(hdi_min))
+        hdi_max = np.arctan2(np.sin(hdi_max), np.cos(hdi_max))
 
-        hpd_intervals = []
-        for interval in intervals_splitted:
-            if interval.size == 0:
-                hpd_intervals.append((bins[0], bins[0]))
-            else:
-                hpd_intervals.append((interval[0], interval[-1]))
+    hpd_intervals = np.array([hdi_min, hdi_max])
 
-        hpd_intervals = np.array(hpd_intervals)
+    return hpd_intervals
 
-    else:
-        if skipna:
-            nans = np.isnan(ary)
-            if not nans.all():
-                ary = ary[~nans]
-        n = len(ary)
 
-        if circular:
-            mean = st.circmean(ary, high=np.pi, low=-np.pi)
-            ary = ary - mean
-            ary = np.arctan2(np.sin(ary), np.cos(ary))
+def _hpd_multimodal(ary, credible_interval, skipna):
+    """Compute hpd if the distribution is multimodal."""
+    ary = ary.flatten()
+    if skipna:
+        ary = ary[~np.isnan(ary)]
 
-        ary = np.sort(ary)
-        interval_idx_inc = int(np.floor(credible_interval * n))
-        n_intervals = n - interval_idx_inc
-        interval_width = ary[interval_idx_inc:] - ary[:n_intervals]
+    if ary.dtype.kind == "f":
+        density, lower, upper = _fast_kde(ary)
+        range_x = upper - lower
+        dx = range_x / len(density)
+        bins = np.linspace(lower, upper, len(density))
+    else:
+        bins = get_bins(ary)
+        _, density, _ = histogram(ary, bins=bins)
+        dx = np.diff(bins)[0]
 
-        if len(interval_width) == 0:
-            raise ValueError("Too few elements for interval calculation. ")
+    density *= dx
 
-        min_idx = np.argmin(interval_width)
-        hdi_min = ary[min_idx]
-        hdi_max = ary[min_idx + interval_idx_inc]
+    idx = np.argsort(-density)
+    intervals = bins[idx][density[idx].cumsum() <= credible_interval]
+    intervals.sort()
 
-        if circular:
-            hdi_min = hdi_min + mean
-            hdi_max = hdi_max + mean
-            hdi_min = np.arctan2(np.sin(hdi_min), np.cos(hdi_min))
-            hdi_max = np.arctan2(np.sin(hdi_max), np.cos(hdi_max))
+    intervals_splitted = np.split(intervals, np.where(np.diff(intervals) >= dx * 1.1)[0] + 1)
 
-        hpd_intervals = np.array([hdi_min, hdi_max])
+    hpd_intervals = []
+    for interval in intervals_splitted:
+        if interval.size == 0:
+            hpd_intervals.append((bins[0], bins[0]))
+        else:
+            hpd_intervals.append((interval[0], interval[-1]))
```

Comment:
On second thought, this could be checked before the loop checking the length of `intervals_splitted`, and if necessary, truncating `intervals_splitted` and issuing the warning

Reply:
What could be the default value for `max_modes`? I am currently using 10.

Reply:
I guess if we think about this in detail and/or run a few tests we should be able to figure out a reasonable default number for this, based on the accuracy/reliability of the method as the number of modes increase. In the meantime 10 sounds good and reminds me of this story 

In 1975, at the age of 99, Leonor Acevedo Borges, the writer’s mother, died. At the wake, a woman expresses her condolences to Borges and comments:
    Poor little Leonetta! To die so close to his 100th birthday! If she had waited a little longer…To what Borges answers:- I see madam that you are devoted to the metric system.

Reply:
Okay. I think I am also one of the devoted ones. 😄